xcm/pallet-xcm/src/tests.rs

```diff
@@ -0,0 +1,180 @@
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::mock::*;
+use frame_support::{assert_noop, assert_ok, traits::Currency};
+use polkadot_parachain::primitives::{AccountIdConversion, Id as ParaId};
+use xcm::{
+	opaque::v0::prelude::*,
+	v0::{Junction, Xcm},
+};
+
+#[test]
+fn send_reserve_asset_deposit_works() {
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = 2 * BaseXcmWeight::get();
+		let sender: MultiLocation =
+			Junction::AccountId32 { network: AnyNetwork::get(), id: ALICE.into() }.into();
+		let message = Xcm::ReserveAssetDeposit {
+			assets: vec![ConcreteFungible { id: Parent.into(), amount }],
+			effects: vec![
+				buy_execution(weight),
+				DepositAsset { assets: vec![All], dest: sender.clone() },
+			],
+		};
+		assert_ok!(XcmPallet::send(Origin::signed(ALICE), RelayLocation::get(), message.clone()));
+		assert_eq!(
+			sent_xcm(),
+			vec![(
+				MultiLocation::Null,
+				RelayedFrom { who: sender.clone(), message: Box::new(message.clone()) }
+			)]
+		);
+		assert_eq!(
+			last_event(),
+			Event::XcmPallet(crate::Event::Sent(sender, RelayLocation::get(), message))
+		);
+	});
+}
+
+#[test]
+fn send_fails_when_xcm_router_blocks() {
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = 2 * BaseXcmWeight::get();
+		let sender: MultiLocation =
+			Junction::AccountId32 { network: AnyNetwork::get(), id: ALICE.into() }.into();
+		let message = Xcm::ReserveAssetDeposit {
+			assets: vec![ConcreteFungible { id: Parent.into(), amount }],
+			effects: vec![
+				buy_execution(weight),
+				DepositAsset { assets: vec![All], dest: sender.clone() },
+			],
+		};
+		assert_noop!(
+			XcmPallet::send(
+				Origin::signed(ALICE),
+				X3(Junction::Parent, Junction::Parent, Junction::Parent),
+				message.clone()
+			),
+			crate::Error::<Test>::SendFailure
+		);
+	});
+}
+
+#[test]
+fn teleport_assets_works() {
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = 2 * BaseXcmWeight::get();
+		assert_eq!(Balances::total_balance(&ALICE), INITIAL_BALANCE);
+		assert_ok!(XcmPallet::teleport_assets(
+			Origin::signed(ALICE),
+			RelayLocation::get(),
+			RelayLocation::get(),
```

Comment:
Why wouldnt this at least send to another user on the chain and you can confirm the balance is sent without fees?

Reply:
We can verify the amount is sent, but we cannot verify the recipient balance increases (effects on destination chain) because the context of these tests is limited to the sending chain.